### PR TITLE
[release-1.20] Fix service entry merge

### PR DESCRIPTION
**Pull request by zirain**
**Please provide a description of this PR:**

fixes: https://github.com/istio/istio/issues/50774

cherry-pick https://github.com/istio/istio/pull/50711

### CODE REVIEW

File: pilot/pkg/model/sidecar.go [hunk 1]
```diff
@@ -355,6 +356,8 @@ func ConvertToSidecarScope(ps *PushContext, sidecarConfig *config.Config, config
 			// Update index as well, so that future reads will merge into the new service
 			foundSvc.svc = copied
 			servicesAdded[foundSvc.svc.Hostname] = foundSvc
+			// update the existing service in the map to the merged one
+			out.servicesByHostname[s.Hostname] = copied
```
Comment (dio) [inline pilot/pkg/model/sidecar.go:360]: Sorry, if we remove this, `TestCreateSidecarScope` (in sidecar_test.go) seems still successful. Do we miss something?

Comment (zirain) [reply pilot/pkg/model/sidecar.go:360]: part of https://github.com/istio/istio/pull/50691

Comment (dio) [reply pilot/pkg/model/sidecar.go:360]: I acknowledge that, but it seems the test in this version doesn't validate the claimed assertion here: https://github.com/istio/istio/pull/50691#discussion_r1581995972.

Comment (zirain) [reply pilot/pkg/model/sidecar.go]: make sense, reverted

**Review by aryan16 — APPROVED** [review]

**Pull request merged**